File: server/scripts/hex_map_generator.py
#!/usr/bin/env python3
"""
Hex Map Geography Generator

Generates a 15x10 hex map with terrain types that cluster naturally.
Each hex considers neighboring terrain to increase odds of similar terrain types.
"""

import random
import json
import sys
from typing import List, Tuple, Optional, Dict

# Default terrain types (fallback if database not available)
DEFAULT_TERRAINS = (
    "plains", "forest", "dark_forest", "hills", "mountains",
    "lake", "marshlands", "quagmire", "ruins"
)

# Terrain compatibility weights - how likely terrain types are to be adjacent
# Higher values = more likely to appear next to each other
TERRAIN_COMPATIBILITY = {
    "plains": {
        "plains": 3.0, "forest": 2.5, "hills": 2.0, "lake": 1.5,
        "dark_forest": 0.8, "mountains": 1.0, "marshlands": 1.2,
        "quagmire": 0.5, "ruins": 1.8
    },
    "forest": {
        "forest": 3.0, "plains": 2.5, "dark_forest": 2.0, "hills": 2.2,
        "mountains": 1.5, "lake": 1.8, "marshlands": 1.0,
        "quagmire": 0.3, "ruins": 1.5
    },
    "dark_forest": {
        "dark_forest": 3.0, "forest": 2.0, "marshlands": 2.5, "quagmire": 2.0,
        "ruins": 2.5, "mountains": 1.5, "hills": 1.2,
        "plains": 0.8, "lake": 0.5
    },
    "hills": {
        "hills": 3.0, "mountains": 2.5, "plains": 2.0, "forest": 2.2,
        "dark_forest": 1.2, "lake": 1.0, "marshlands": 0.8,
        "quagmire": 0.3, "ruins": 2.0
    },
    "mountains": {
        "mountains": 3.0, "hills": 2.5, "forest": 1.5, "dark_forest": 1.5,
        "lake": 1.2, "plains": 1.0, "marshlands": 0.3,
        "quagmire": 0.1, "ruins": 1.8
    },
    "lake": {
        "lake": 2.5, "plains": 1.5, "forest": 1.8, "marshlands": 2.5,
        "quagmire": 1.8, "hills": 1.0, "mountains": 1.2,
        "dark_forest": 0.5, "ruins": 1.0
    },
    "marshlands": {
        "marshlands": 3.0, "quagmire": 2.8, "lake": 2.5, "dark_forest": 2.5,
        "forest": 1.0, "plains": 1.2, "hills": 0.8,
        "mountains": 0.3, "ruins": 1.5
    },
    "quagmire": {
        "quagmire": 3.0, "marshlands": 2.8, "dark_forest": 2.0, "lake": 1.8,
        "ruins": 2.0, "forest": 0.3, "plains": 0.5,
        "hills": 0.3, "mountains": 0.1
    },
    "ruins": {
        "ruins": 2.0, "dark_forest": 2.5, "quagmire": 2.0, "hills": 2.0,
        "mountains": 1.8, "plains": 1.8, "forest": 1.5,
        "marshlands": 1.5, "lake": 1.0
    }
}

class HexMapGenerator:
    def __init__(self, width: int = 15, height: int = 10, terrains: Tuple[str, ...] = None):
        self.width = width
        self.height = height
        self.terrains = terrains or DEFAULT_TERRAINS
        self.map_grid: List[List[str]] = []

        # Validate terrain compatibility matrix
        self._validate_compatibility_matrix()

    def _validate_compatibility_matrix(self):
        """Ensure all terrain types have compatibility entries"""
        for terrain in self.terrains:
            if terrain not in TERRAIN_COMPATIBILITY:
                # Create default compatibility for missing terrains
                TERRAIN_COMPATIBILITY[terrain] = {t: 1.0 for t in self.terrains}

        # Ensure all terrain types are referenced in each compatibility dict
        for terrain in self.terrains:
            for other_terrain in self.terrains:
                if other_terrain not in TERRAIN_COMPATIBILITY[terrain]:
                    TERRAIN_COMPATIBILITY[terrain][other_terrain] = 1.0

    def get_neighbors(self, row: int, col: int) -> List[str]:
        """Get terrain types of neighboring hexes (left and above)"""
        neighbors = []

        # Left neighbor
        if col > 0:
            neighbors.append(self.map_grid[row][col - 1])

        # Top neighbor
        if row > 0:
            neighbors.append(self.map_grid[row - 1][col])

        return neighbors

    def calculate_terrain_weights(self, neighbors: List[str]) -> Dict[str, float]:
        """Calculate weighted probabilities for terrain types based on neighbors"""
        if not neighbors:
            # No neighbors - equal probability for all terrains
            return {terrain: 1.0 for terrain in self.terrains}

        weights = {}
        for terrain in self.terrains:
            # Start with base weight
            weight = 1.0

            # Multiply by compatibility with each neighbor
            for neighbor in neighbors:
                compatibility = TERRAIN_COMPATIBILITY.get(neighbor, {}).get(terrain, 1.0)
                weight *= compatibility

            # Average the influence if multiple neighbors
            if len(neighbors) > 1:
                weight = weight ** (1.0 / len(neighbors))

            weights[terrain] = weight

        return weights

    def weighted_terrain_choice(self, weights: Dict[str, float]) -> str:
        """Choose a terrain type based on weighted probabilities"""
        terrain_list = list(weights.keys())
        weight_list = list(weights.values())

        # Normalize weights to sum to 1
        total_weight = sum(weight_list)
        if total_weight > 0:
            weight_list = [w / total_weight for w in weight_list]
        else:
            # Fallback to equal weights
            weight_list = [1.0 / len(terrain_list)] * len(terrain_list)

        # Manual weighted choice for better compatibility
        rand_val = random.random()
        cumulative = 0.0

        for i, weight in enumerate(weight_list):
            cumulative += weight
            if rand_val <= cumulative:
                return terrain_list[i]

        # Fallback - return last terrain
        return terrain_list[-1] if terrain_list else self.terrains[0]

    def generate_map(self, seed: Optional[int] = None) -> List[List[str]]:
        """Generate the complete hex map"""
        if seed is not None:
            random.seed(seed)

        self.map_grid = []

        for row in range(self.height):
            current_row = []

            for col in range(self.width):
                # Get neighboring terrain types
                neighbors = self.get_neighbors(row, col)

                # Calculate weights based on neighbors
                weights = self.calculate_terrain_weights(neighbors)

                # Choose terrain type
                terrain = self.weighted_terrain_choice(weights)
                current_row.append(terrain)

            self.map_grid.append(current_row)

        return self.map_grid

    def map_to_dict(self) -> Dict:
        """Convert map to dictionary format for JSON output"""
        result = {
            "width": self.width,
            "height": self.height,
            "terrains": list(self.terrains),
            "hexes": []
        }

        for row in range(self.height):
            for col in range(self.width):
                result["hexes"].append({
                    "row": row,
                    "col": col,
                    "terrain": self.map_grid[row][col],
                    "id": f"hex_{row}_{col}"
                })

        return result

    def print_map_ascii(self) -> str:
        """Generate ASCII representation of the map for debugging"""
        terrain_symbols = {
            "plains": "P", "forest": "F", "dark_forest": "D", "hills": "H",
            "mountains": "M", "lake": "L", "marshlands": "W", "quagmire": "Q", "ruins": "R"
        }

        result = []
        result.append(f"Hex Map ({self.width}x{self.height}):")
        result.append("=" * (self.width * 2 + 5))

        for row in range(self.height):
            row_str = f"{row:2d}| "
            for col in range(self.width):
                terrain = self.map_grid[row][col]
                symbol = terrain_symbols.get(terrain, "?")
                row_str += symbol + " "
            result.append(row_str)

        # Add column numbers
        col_header = "   "
        for col in range(self.width):
            col_header += f"{col % 10} "
        result.append("=" * (self.width * 2 + 5))
        result.append(col_header)

        return "\n".join(result)


def main():
    """Main function for command line usage"""
    # Parse command line arguments
    if len(sys.argv) > 1:
        command = sys.argv[1]

        if command == "generate":
            # Get optional parameters
            width = int(sys.argv[2]) if len(sys.argv) > 2 else 15
            height = int(sys.argv[3]) if len(sys.argv) > 3 else 10
            seed = int(sys.argv[4]) if len(sys.argv) > 4 else None

            # Generate map
            generator = HexMapGenerator(width, height, DEFAULT_TERRAINS)
            hex_map = generator.generate_map(seed)

            # Output JSON
            result = generator.map_to_dict()
            result["success"] = True
            print(json.dumps(result))

        elif command == "test":
            # Generate test map and show ASCII
            generator = HexMapGenerator(15, 10, DEFAULT_TERRAINS)
            hex_map = generator.generate_map(42)  # Fixed seed for testing

            print(generator.print_map_ascii())
            print("\nJSON Preview:")
            result = generator.map_to_dict()
            result["success"] = True
            print(json.dumps(result, indent=2)[:500] + "...")

        elif command == "terrains":
            # Return available terrain types
            result = {
                "success": True,
                "terrains": list(DEFAULT_TERRAINS),
                "compatibility_matrix": TERRAIN_COMPATIBILITY
            }
            print(json.dumps(result))

        else:
            # Unknown command
            result = {
                "success": False,
                "error": f"Unknown command: {command}",
                "available_commands": ["generate", "test", "terrains"]
            }
            print(json.dumps(result))

    else:
        # Default behavior - generate a standard map
        generator = HexMapGenerator()
        hex_map = generator.generate_map()
        result = generator.map_to_dict()
        result["success"] = True
        print(json.dumps(result))


if __name__ == "__main__":
    main()

File: server/lib/hex-map-generator.ts
/**
 * Hex Map Geography Generator (TypeScript)
 *
 * Generates a 15x10 hex map with terrain types that cluster naturally.
 * Each hex considers neighboring terrain to increase odds of similar terrain types.
 */

// Default terrain types (fallback if database not available)
export const DEFAULT_TERRAINS = [
  "plains",
  "forest",
  "dark_forest",
  "hills",
  "mountains",
  "lake",
  "marshlands",
  "ruins",
] as const;

export type TerrainType = (typeof DEFAULT_TERRAINS)[number];

// Terrain compatibility weights - how likely terrain types are to be adjacent
// Higher values = more likely to appear next to each other
export const TERRAIN_COMPATIBILITY: Record<
  TerrainType,
  Record<TerrainType, number>
> = {
  plains: {
    plains: 5.0,
    forest: 3.0,
    hills: 2.5,
    lake: 2.0,
    dark_forest: 0.5,
    mountains: 1.0,
    marshlands: 1.5,
    ruins: 2.0,
  },
  forest: {
    forest: 5.0,
    plains: 3.0,
    dark_forest: 3.5,
    hills: 2.8,
    mountains: 2.0,
    lake: 2.2,
    marshlands: 1.5,
    ruins: 1.8,
  },
  dark_forest: {
    dark_forest: 5.0,
    forest: 3.0,
    marshlands: 3.5,
    ruins: 3.5,
    mountains: 2.0,
    hills: 1.5,
    plains: 0.5,
    lake: 0.3,
  },
  hills: {
    hills: 5.0,
    mountains: 4.0,
    plains: 3.0,
    forest: 2.8,
    dark_forest: 1.5,
    lake: 1.2,
    marshlands: 0.5,
    ruins: 2.5,
  },
  mountains: {
    mountains: 5.0,
    hills: 4.0,
    forest: 2.0,
    dark_forest: 2.0,
    lake: 1.5,
    plains: 1.2,
    marshlands: 0.2,
    ruins: 2.2,
  },
  lake: {
    lake: 4.0,
    plains: 2.0,
    forest: 2.5,
    marshlands: 3.5,
    hills: 1.2,
    mountains: 1.5,
    dark_forest: 0.3,
    ruins: 1.2,
  },
  marshlands: {
    marshlands: 5.0,
    lake: 3.5,
    dark_forest: 3.5,
    forest: 1.2,
    plains: 1.5,
    hills: 0.5,
    mountains: 0.2,
    ruins: 2.0,
  },
  ruins: {
    ruins: 3.0,
    dark_forest: 3.5,
    hills: 2.0,
    mountains: 1.8,
    plains: 1.8,
    forest: 1.5,
    marshlands: 1.5,
    lake: 1.0,
  },
};

export interface HexCell {
  row: number;
  col: number;
  terrain: string;
  id: string;
}

export interface HexMapResult {
  success: boolean;
  width?: number;
  height?: number;
  terrains?: string[];
  hexes?: HexCell[];
  error?: string;
}

export class HexMapGenerator {
  private width: number;
  private height: number;
  private terrains: string[];
  private mapGrid: string[][] = [];

  constructor(width: number = 15, height: number = 10, terrains?: string[]) {
    this.width = width;
    this.height = height;
    this.terrains = terrains || [...DEFAULT_TERRAINS];

    // Validate terrain compatibility matrix
    this.validateCompatibilityMatrix();
  }

  private validateCompatibilityMatrix(): void {
    // Ensure all terrain types have compatibility entries
    for (const terrain of this.terrains) {
      if (!(terrain in TERRAIN_COMPATIBILITY)) {
        // Create default compatibility for missing terrains
        (TERRAIN_COMPATIBILITY as any)[terrain] = {};
        for (const t of this.terrains) {
          (TERRAIN_COMPATIBILITY as any)[terrain][t] = 1.0;
        }
      }
    }

    // Ensure all terrain types are referenced in each compatibility dict
    for (const terrain of this.terrains) {
      for (const otherTerrain of this.terrains) {
        if (!(otherTerrain in (TERRAIN_COMPATIBILITY as any)[terrain])) {
          (TERRAIN_COMPATIBILITY as any)[terrain][otherTerrain] = 1.0;
        }
      }
    }
  }

  private getNeighbors(row: number, col: number): string[] {
    const neighbors: string[] = [];

    // In a hex grid, we have 6 neighbors, but since we're generating row by row,
    // we can only consider already-generated neighbors (above and to the left)

    // Top neighbor
    if (row > 0) {
      neighbors.push(this.mapGrid[row - 1][col]);
    }

    // Top-left neighbor (offset for hex grid)
    if (row > 0 && col > 0) {
      neighbors.push(this.mapGrid[row - 1][col - 1]);
    }

    // Top-right neighbor (offset for hex grid)
    if (row > 0 && col < this.width - 1) {
      // For hex grid, odd/even rows have different offsets
      const offset = col % 2 === 0 ? 0 : 1;
      if (col + offset < this.width) {
        neighbors.push(this.mapGrid[row - 1][col + offset]);
      }
    }

    // Left neighbor - CRITICAL for ruins constraint
    if (col > 0) {
      neighbors.push(this.mapGrid[row][col - 1]);
    }

    // For clustering effect, add key neighbors multiple times
    // But only add them once for ruins constraint to work properly
    if (row > 0) {
      neighbors.push(this.mapGrid[row - 1][col]); // Add top neighbor again for clustering
    }

    return neighbors;
  }

  private calculateTerrainWeights(neighbors: string[]): Record<string, number> {
    if (neighbors.length === 0) {
      // No neighbors - equal probability for all terrains
      const weights: Record<string, number> = {};
      for (const terrain of this.terrains) {
        weights[terrain] = 1.0;
      }
      return weights;
    }

    const weights: Record<string, number> = {};
    for (const terrain of this.terrains) {
      try {
        // Start with base weight
        let weight = 1.0;

        // Special rule: Ruins can never be adjacent to other ruins
        if (terrain === "ruins") {
          const hasRuinsNeighbor = neighbors.some(
            (neighbor) => neighbor === "ruins",
          );
          if (hasRuinsNeighbor) {
            weight = 0.0; // Completely prevent ruins next to ruins
          }
        }

        // Skip compatibility calculation if weight is already 0
        if (weight > 0) {
          // Multiply by compatibility with each neighbor
          for (const neighbor of neighbors) {
            const compatibility =
              (TERRAIN_COMPATIBILITY as any)[neighbor]?.[terrain] ?? 1.0;
            // Square the compatibility to increase clustering bias
            weight *= Math.pow(compatibility, 1.5);
          }

          // Use geometric mean but with less averaging to maintain clustering
          if (neighbors.length > 1) {
            weight = Math.pow(weight, 0.8 / neighbors.length);
          }
        }

        weights[terrain] = weight;
      } catch (error) {
        // Fallback to equal weight if error
        weights[terrain] = 1.0;
      }
    }

    return weights;
  }

  private weightedTerrainChoice(weights: Record<string, number>): string {
    const terrainList = Object.keys(weights);
    const weightList = Object.values(weights);

    // Normalize weights to sum to 1
    const totalWeight = weightList.reduce((sum, w) => sum + w, 0);
    let normalizedWeights: number[];

    if (totalWeight > 0) {
      normalizedWeights = weightList.map((w) => w / totalWeight);
    } else {
      // Fallback to equal weights
      normalizedWeights = new Array(terrainList.length).fill(
        1.0 / terrainList.length,
      );
    }

    // Manual weighted choice
    const randVal = Math.random();
    let cumulative = 0.0;

    for (let i = 0; i < normalizedWeights.length; i++) {
      cumulative += normalizedWeights[i];
      if (randVal <= cumulative) {
        return terrainList[i];
      }
    }

    // Fallback - return last terrain
    return terrainList[terrainList.length - 1] || this.terrains[0];
  }

  private getAllNeighbors(
    row: number,
    col: number,
  ): Array<{ row: number; col: number }> {
    const neighbors: Array<{ row: number; col: number }> = [];

    // All 6 hexagonal neighbors
    const offsets = [
      [-1, 0], // Top
      [-1, -1], // Top-left
      [-1, 1], // Top-right
      [0, -1], // Left
      [0, 1], // Right
      [1, 0], // Bottom
      [1, -1], // Bottom-left
      [1, 1], // Bottom-right
    ];

    for (const [dr, dc] of offsets) {
      const newRow = row + dr;
      const newCol = col + dc;

      if (
        newRow >= 0 &&
        newRow < this.height &&
        newCol >= 0 &&
        newCol < this.width
      ) {
        neighbors.push({ row: newRow, col: newCol });
      }
    }

    return neighbors;
  }

  private fixIsolatedTiles(): void {
    // Remove isolated single tiles by converting them to neighbor terrain
    for (let row = 0; row < this.height; row++) {
      for (let col = 0; col < this.width; col++) {
        const currentTerrain = this.mapGrid[row][col];
        const neighbors = this.getAllNeighbors(row, col);

        // Count neighbors with same terrain
        let sameTerrainCount = 0;
        const differentTerrains: string[] = [];

        for (const neighbor of neighbors) {
          const neighborTerrain = this.mapGrid[neighbor.row][neighbor.col];
          if (neighborTerrain === currentTerrain) {
            sameTerrainCount++;
          } else {
            differentTerrains.push(neighborTerrain);
          }
        }

        // If no same-terrain neighbors, convert to most common neighbor terrain
        if (sameTerrainCount === 0 && differentTerrains.length > 0) {
          const terrainCounts: Record<string, number> = {};
          for (const terrain of differentTerrains) {
            terrainCounts[terrain] = (terrainCounts[terrain] || 0) + 1;
          }

          let mostCommon = differentTerrains[0];
          let maxCount = 0;
          for (const [terrain, count] of Object.entries(terrainCounts)) {
            if (count > maxCount) {
              maxCount = count;
              mostCommon = terrain;
            }
          }

          this.mapGrid[row][col] = mostCommon;
        }
      }
    }
  }

  private fixAdjacentRuins(): void {
    // Fix any ruins that ended up adjacent to other ruins
    for (let row = 0; row < this.height; row++) {
      for (let col = 0; col < this.width; col++) {
        if (this.mapGrid[row][col] === "ruins") {
          const neighbors = this.getAllNeighbors(row, col);

          // Check if any neighbors are also ruins
          const hasRuinsNeighbor = neighbors.some(
            (neighbor) => this.mapGrid[neighbor.row][neighbor.col] === "ruins",
          );

          if (hasRuinsNeighbor) {
            // Convert this ruins to the most common non-ruins neighbor terrain
            const nonRuinsNeighbors = neighbors
              .map((neighbor) => this.mapGrid[neighbor.row][neighbor.col])
              .filter((terrain) => terrain !== "ruins");

            if (nonRuinsNeighbors.length > 0) {
              const terrainCounts: Record<string, number> = {};
              for (const terrain of nonRuinsNeighbors) {
                terrainCounts[terrain] = (terrainCounts[terrain] || 0) + 1;
              }

              let mostCommon = nonRuinsNeighbors[0];
              let maxCount = 0;
              for (const [terrain, count] of Object.entries(terrainCounts)) {
                if (count > maxCount) {
                  maxCount = count;
                  mostCommon = terrain;
                }
              }

              this.mapGrid[row][col] = mostCommon;
            }
          }
        }
      }
    }
  }

  private findConnectedGroup(
    startRow: number,
    startCol: number,
    terrain: string,
    visited: boolean[][],
  ): Array<{ row: number; col: number }> {
    const group: Array<{ row: number; col: number }> = [];
    const stack = [{ row: startRow, col: startCol }];

    while (stack.length > 0) {
      const current = stack.pop()!;
      const { row, col } = current;

      if (visited[row][col] || this.mapGrid[row][col] !== terrain) {
        continue;
      }

      visited[row][col] = true;
      group.push(current);

      // Add unvisited same-terrain neighbors to stack
      const neighbors = this.getAllNeighbors(row, col);
      for (const neighbor of neighbors) {
        if (
          !visited[neighbor.row][neighbor.col] &&
          this.mapGrid[neighbor.row][neighbor.col] === terrain
        ) {
          stack.push(neighbor);
        }
      }
    }

    return group;
  }

  private ensureMinimumGroups(): void {
    // Find all connected groups
    const visited: boolean[][] = Array(this.height)
      .fill(null)
      .map(() => Array(this.width).fill(false));
    const groups: Array<{
      terrain: string;
      positions: Array<{ row: number; col: number }>;
    }> = [];

    for (let row = 0; row < this.height; row++) {
      for (let col = 0; col < this.width; col++) {
        if (!visited[row][col]) {
          const terrain = this.mapGrid[row][col];
          const group = this.findConnectedGroup(row, col, terrain, visited);
          if (group.length >= 3) {
            groups.push({ terrain, positions: group });
          }
        }
      }
    }

    // If we don't have at least 3 groups of 3+, create them
    if (groups.length < 3) {
      const targetTerrains = this.terrains.slice(0, 3); // Use first 3 terrain types

      for (let i = groups.length; i < 3; i++) {
        const terrain = targetTerrains[i % targetTerrains.length];

        // Find a good spot to place a 3+ group
        let placed = false;
        for (let attempts = 0; attempts < 20 && !placed; attempts++) {
          const centerRow = Math.floor(Math.random() * this.height);
          const centerCol = Math.floor(Math.random() * this.width);

          // Try to place a 3-hex group around this center
          const groupPositions = [
            { row: centerRow, col: centerCol },
            { row: centerRow, col: Math.max(0, centerCol - 1) },
            { row: Math.max(0, centerRow - 1), col: centerCol },
          ].filter((pos) => pos.row < this.height && pos.col < this.width);

          if (groupPositions.length >= 3) {
            for (const pos of groupPositions) {
              this.mapGrid[pos.row][pos.col] = terrain;
            }
            placed = true;
          }
        }
      }
    }
  }

  public generateMap(seed?: number): string[][] {
    try {
      if (seed !== undefined) {
        // Simple seed-based random number generator
        this.seedRandom(seed);
      }

      this.mapGrid = [];

      for (let row = 0; row < this.height; row++) {
        const currentRow: string[] = [];

        for (let col = 0; col < this.width; col++) {
          try {
            // Get neighboring terrain types
            const neighbors = this.getNeighbors(row, col);

            // Calculate weights based on neighbors
            const weights = this.calculateTerrainWeights(neighbors);

            // Choose terrain type
            const terrain = this.weightedTerrainChoice(weights);
            currentRow.push(terrain);
          } catch (error) {
            // Fallback to random terrain
            const terrain =
              this.terrains[Math.floor(Math.random() * this.terrains.length)];
            currentRow.push(terrain);
          }
        }

        this.mapGrid.push(currentRow);
      }

      // Post-processing for better clustering
      this.fixIsolatedTiles();
      this.fixIsolatedTiles(); // Run twice for better results
      this.ensureMinimumGroups();

      return this.mapGrid;
    } catch (error) {
      // Generate simple fallback map
      this.mapGrid = [];
      for (let row = 0; row < this.height; row++) {
        const currentRow: string[] = [];
        for (let col = 0; col < this.width; col++) {
          const terrain =
            this.terrains[Math.floor(Math.random() * this.terrains.length)];
          currentRow.push(terrain);
        }
        this.mapGrid.push(currentRow);
      }
      return this.mapGrid;
    }
  }

  public mapToDict(): HexMapResult {
    const result: HexMapResult = {
      success: true,
      width: this.width,
      height: this.height,
      terrains: [...this.terrains],
      hexes: [],
    };

    for (let row = 0; row < this.height; row++) {
      for (let col = 0; col < this.width; col++) {
        result.hexes!.push({
          row,
          col,
          terrain: this.mapGrid[row][col],
          id: `hex_${row}_${col}`,
        });
      }
    }

    return result;
  }

  public printMapAscii(): string {
    const terrainSymbols: Record<string, string> = {
      plains: "P",
      forest: "F",
      dark_forest: "D",
      hills: "H",
      mountains: "M",
      lake: "L",
      marshlands: "W",
      ruins: "R",
    };

    const result: string[] = [];
    result.push(`Hex Map (${this.width}x${this.height}):`);
    result.push("=".repeat(this.width * 2 + 5));

    for (let row = 0; row < this.height; row++) {
      let rowStr = `${row.toString().padStart(2)}| `;
      for (let col = 0; col < this.width; col++) {
        const terrain = this.mapGrid[row][col];
        const symbol = terrainSymbols[terrain] || "?";
        rowStr += symbol + " ";
      }
      result.push(rowStr);
    }

    // Add column numbers
    let colHeader = "   ";
    for (let col = 0; col < this.width; col++) {
      colHeader += `${col % 10} `;
    }
    result.push("=".repeat(this.width * 2 + 5));
    result.push(colHeader);

    return result.join("\n");
  }

  private seedRandom(seed: number): void {
    // Simple seeded random number generator
    let currentSeed = seed;
    Math.random = () => {
      currentSeed = (currentSeed * 9301 + 49297) % 233280;
      return currentSeed / 233280;
    };
  }
}

// Utility functions for API endpoints
export function generateHexMap(
  width: number = 15,
  height: number = 10,
  seed?: number,
): HexMapResult {
  try {
    const generator = new HexMapGenerator(width, height, [...DEFAULT_TERRAINS]);
    generator.generateMap(seed);
    const result = generator.mapToDict();
    return result;
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

export function getTerrainTypes(): HexMapResult {
  return {
    success: true,
    terrains: [...DEFAULT_TERRAINS],
    compatibility_matrix: TERRAIN_COMPATIBILITY,
  } as any;
}

export function generateTestMap(): {
  success: boolean;
  ascii_output?: string;
  error?: string;
} {
  try {
    const generator = new HexMapGenerator(15, 10, [...DEFAULT_TERRAINS]);
    generator.generateMap(42); // Fixed seed for testing
    const ascii = generator.printMapAscii();

    return {
      success: true,
      ascii_output: ascii,
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}
